File: mobile_client/lib/services/cosmetics_service.dart
import 'package:flutter/foundation.dart';
import 'package:mobile_client/services/auth_service.dart';

class CosmeticBackground {
  final String id;
  final String name;
  /// Blue theme variant (aligned with the Angular `blueValue`).
  final String blueAssetPath;
  /// Red theme variant (Angular `redValue`).
  final String redAssetPath;
  final int price;

  const CosmeticBackground({
    required this.id,
    required this.name,
    required this.blueAssetPath,
    required this.redAssetPath,
    required this.price,
  });

  String assetForTheme(bool isBlue) =>
      isBlue ? blueAssetPath : redAssetPath;
}

class FeaturedAvatar {
  final String id;
  final String name;
  final int price;
  final String icon;
  final String? animation;

  const FeaturedAvatar({
    required this.id,
    required this.name,
    required this.price,
    required this.icon,
    this.animation,
  });
}

class CosmeticMusic {
  final String id;
  final String name;
  final String cover;
  final int price;

  const CosmeticMusic({
    required this.id,
    required this.name,
    required this.cover,
    required this.price,
  });
}

class CosmeticsService extends ChangeNotifier {
  final AuthService authService;

  CosmeticsService({required this.authService});

  static const List<CosmeticBackground> backgrounds = [
    CosmeticBackground(
      id: 'background-default',
      name: 'shop_page.background.default',
      blueAssetPath: 'assets/backgrounds/background-default.png',
      redAssetPath: 'assets/backgrounds/background-default.png',
      price: 0,
    ),
    CosmeticBackground(
      id: 'background-1',
      name: 'shop_page.background.background_1',
      blueAssetPath: 'assets/backgrounds/eggcited_blue.gif',
      redAssetPath: 'assets/backgrounds/eggcited_red.gif',
      price: 400,
    ),
    CosmeticBackground(
      id: 'background-2',
      name: 'shop_page.background.background_2',
      blueAssetPath: 'assets/backgrounds/zero-duck-given_blue.gif',
      redAssetPath: 'assets/backgrounds/zero-duck-given_red.gif',
      price: 600,
    ),
    CosmeticBackground(
      id: 'background-3',
      name: 'shop_page.background.background_3',
      blueAssetPath: 'assets/backgrounds/champions-brew_blue.gif',
      redAssetPath: 'assets/backgrounds/champions-brew_red.gif',
      price: 800,
    ),
  ];

  static const List<FeaturedAvatar> featuredAvatars = [
    FeaturedAvatar(
      id: 'specter',
      name: 'Specter',
      price: 500,
      icon: 'assets/avatar_icon/specter_icon.png',
      animation: 'assets/avatar_gif/specter.gif',
    ),
    FeaturedAvatar(
      id: 'titan',
      name: 'Titan',
      price: 750,
      icon: 'assets/avatar_icon/titan_icon.png',
      animation: 'assets/avatar_gif/titan.gif',
    ),
    FeaturedAvatar(
      id: 'whiplash',
      name: 'Whiplash',
      price: 1000,
      icon: 'assets/avatar_icon/whiplash_icon.png',
      animation: 'assets/avatar_gif/whiplash.gif',
    ),
    FeaturedAvatar(
      id: 'yang',
      name: 'Yang',
      price: 1500,
      icon: 'assets/avatar_icon/yang_icon.png',
      animation: 'assets/avatar_gif/yang.gif',
    ),
  ];

  static const List<CosmeticMusic> musics = [
    CosmeticMusic(
      id: 'music-default',
      name: 'music_name.main_theme',
      cover: 'assets/music-covers/music-default.png',
      price: 0,
    ),
    CosmeticMusic(
      id: 'music-1',
      name: 'music_name.epic',
      cover: 'assets/music-covers/music-1.png',
      price: 300,
    ),
    CosmeticMusic(
      id: 'music-2',
      name: 'music_name.adventure',
      cover: 'assets/music-covers/music-2.png',
      price: 500,
    ),
    CosmeticMusic(
      id: 'music-3',
      name: 'music_name.mystic',
      cover: 'assets/music-covers/music-3.png',
      price: 700,
    ),
  ];

  List<CosmeticBackground> get shopBackgrounds =>
      backgrounds.where((b) => b.id != 'background-default').toList();

  List<FeaturedAvatar> get shopAvatars => featuredAvatars;

  List<CosmeticMusic> get shopMusics =>
      musics.where((m) => m.id != 'music-default').toList();

  List<CosmeticBackground> getOwnedBackgrounds(List<String> ownedIds) {
    return backgrounds.where((b) => ownedIds.contains(b.id)).toList();
  }

  List<FeaturedAvatar> getOwnedAvatars(List<String> ownedNames) {
    return featuredAvatars
        .where((a) => ownedNames.contains(a.name) || ownedNames.contains(a.id))
        .toList();
  }

  List<CosmeticMusic> getOwnedMusics(List<String> ownedIds) {
    return musics.where((m) => ownedIds.contains(m.id)).toList();
  }

  bool isBackgroundOwned(String bgId) {
    final user = authService.currentUser;
    return user?.ownedBackgrounds.contains(bgId) ?? (bgId == 'background-default');
  }

  bool isAvatarOwned(String avatarName) {
    final user = authService.currentUser;
    return user?.ownedAvatars.contains(avatarName) ?? false;
  }

  bool isMusicOwned(String musicId) {
    final user = authService.currentUser;
    return user?.ownedMusics.contains(musicId) ?? (musicId == 'music-default');
  }

  bool canAfford(int price) {
    return (authService.currentUser?.virtualCurrency ?? 0) >= price;
  }

  int get userCurrency => authService.currentUser?.virtualCurrency ?? 0;

  Future<bool> purchaseBackground(CosmeticBackground bg) async {
    if (isBackgroundOwned(bg.id) || !canAfford(bg.price)) return false;
    final result = await authService.purchaseBackground(bg.id, bg.price);
    if (result) notifyListeners();
    return result;
  }

  Future<bool> purchaseAvatar(FeaturedAvatar avatar) async {
    if (isAvatarOwned(avatar.name) || !canAfford(avatar.price)) return false;
    final result = await authService.purchaseAvatar(avatar.name, avatar.price);
    if (result) notifyListeners();
    return result;
  }

  Future<bool> purchaseMusic(CosmeticMusic music) async {
    if (isMusicOwned(music.id) || !canAfford(music.price)) return false;
    final result = await authService.purchaseMusic(music.id, music.price);
    if (result) notifyListeners();
    return result;
  }
}

File: client/src/app/services/cosmetics/cosmetics.service.ts
import { Injectable } from '@angular/core';
import { BACKGROUND_DEFAULT_ID, MUSIC_DEFAULT_ID, MUSIC_OFF_ID } from '@app/constants/cosmetics.constants';
import { AuthService } from '@app/services/auth-service/auth-service.service';
import { CustomizationService } from '@app/services/customization-service/customization.service';
import { BehaviorSubject, Observable, combineLatest, distinctUntilChanged, map } from 'rxjs';

export interface Background {
    id: string;
    name: string;
    type: 'image';
    blueValue: string;
    redValue: string;
    price: number;
}

export interface FeaturedAvatar {
    id: string;
    name: string;
    price: number;
    icon: string;
    animation?: string;
}

export interface Music {
    id: string;
    name: string;
    path: string;
    cover: string;
    price: number;
}

@Injectable({
    providedIn: 'root',
})
export class CosmeticsService {
    backgrounds: Background[] = [
        {
            id: BACKGROUND_DEFAULT_ID,
            name: 'shop_page.background.default',
            type: 'image',
            blueValue: 'url(assets/backgrounds/background-default.png)',
            redValue: 'url(assets/backgrounds/background-default.png)',
            price: 0,
        },
        {
            id: 'background-1',
            name: 'shop_page.background.background_1',
            type: 'image',
            blueValue: 'url(assets/backgrounds/eggcited_blue.gif)',
            redValue: 'url(assets/backgrounds/eggcited_red.gif)',
            price: 400,
        },
        {
            id: 'background-2',
            name: 'shop_page.background.background_2',
            type: 'image',
            blueValue: 'url(assets/backgrounds/zero-duck-given_blue.gif)',
            redValue: 'url(assets/backgrounds/zero-duck-given_red.gif)',
            price: 600,
        },
        {
            id: 'background-3',
            name: 'shop_page.background.background_3',
            type: 'image',
            blueValue: 'url(assets/backgrounds/champions-brew_blue.gif)',
            redValue: 'url(assets/backgrounds/champions-brew_red.gif)',
            price: 800,
        },
    ];

    featuredAvatars: FeaturedAvatar[] = [
        { id: 'specter', name: 'Specter', price: 500, icon: 'assets/avatar_icon/specter_icon.png', animation: 'assets/avatar_gif/specter.gif' },
        { id: 'titan', name: 'Titan', price: 750, icon: 'assets/avatar_icon/titan_icon.png', animation: 'assets/avatar_gif/titan.gif' },
        { id: 'whiplash', name: 'Whiplash', price: 1000, icon: 'assets/avatar_icon/whiplash_icon.png', animation: 'assets/avatar_gif/whiplash.gif' },
        { id: 'yang', name: 'Yang', price: 1500, icon: 'assets/avatar_icon/yang_icon.png', animation: 'assets/avatar_gif/yang.gif' },
    ];

    musics: Music[] = [
        {
            id: MUSIC_DEFAULT_ID,
            name: 'music_name.main_theme',
            path: 'assets/audio/main-theme.mp3',
            cover: 'assets/music-covers/music-default.png',
            price: 0,
        },
        {
            id: 'music-1',
            name: 'music_name.epic',
            path: 'assets/audio/music-1.mp3',
            cover: 'assets/music-covers/music-1.png',
            price: 300,
        },
        {
            id: 'music-2',
            name: 'music_name.adventure',
            path: 'assets/audio/music-2.mp3',
            cover: 'assets/music-covers/music-2.png',
            price: 500,
        },
        {
            id: 'music-3',
            name: 'music_name.mystic',
            path: 'assets/audio/music-3.mp3',
            cover: 'assets/music-covers/music-3.png',
            price: 700,
        },
    ];

    currentBackgroundStyle$!: Observable<string>;
    private readonly selectedBackgroundSubject = new BehaviorSubject<string>(BACKGROUND_DEFAULT_ID);
    private readonly selectedMusicSubject = new BehaviorSubject<string>(MUSIC_DEFAULT_ID);

    constructor(
        private readonly authService: AuthService,
        private readonly customizationService: CustomizationService,
    ) {
        this.currentBackgroundStyle$ = combineLatest([
            this.selectedBackgroundSubject.pipe(distinctUntilChanged()),
            this.customizationService.theme$,
        ]).pipe(map(([bgId]) => this.getBackgroundStyle(bgId)));

        this.authService.userProfile$.subscribe((profile) => {
            if (profile) {
                this.selectedBackgroundSubject.next(profile.selectedBackground || BACKGROUND_DEFAULT_ID);
                this.selectedMusicSubject.next(profile.selectedMusic || MUSIC_DEFAULT_ID);
            } else {
                this.selectedMusicSubject.next(MUSIC_OFF_ID);
                this.selectedBackgroundSubject.next(BACKGROUND_DEFAULT_ID);
            }
        });
    }

    get selectedBackground$() {
        return this.selectedBackgroundSubject.pipe(distinctUntilChanged());
    }

    get selectedMusic$() {
        return this.selectedMusicSubject.pipe(distinctUntilChanged());
    }

    setBackground(bgId: string): void {
        this.authService.updateAccountMongoDB({ selectedBackground: bgId }).subscribe({
            next: (updatedProfile) => {
                this.authService.refreshUserProfile(updatedProfile);
            },
        });
    }

    getBackgroundStyle(bgId: string): string {
        const background = this.backgrounds.find((item) => item.id === bgId) || this.backgrounds[0];
        return this.customizationService.currentTheme === 'red-theme' ? background.redValue : background.blueValue;
    }

    getOwnedBackgrounds(ownedIds: string[]): Background[] {
        return this.backgrounds.filter((background) => ownedIds.includes(background.id));
    }

    getShopBackgrounds(): Background[] {
        return this.backgrounds.filter((background) => background.id !== BACKGROUND_DEFAULT_ID);
    }

    getShopAvatars(): FeaturedAvatar[] {
        return this.featuredAvatars;
    }

    getOwnedAvatars(ownedNames: string[]): FeaturedAvatar[] {
        return this.featuredAvatars.filter((avatar) => ownedNames.includes(avatar.name) || ownedNames.includes(avatar.id));
    }

    setMusic(musicId: string): void {
        this.authService.updateAccountMongoDB({ selectedMusic: musicId }).subscribe({
            next: (updatedProfile) => {
                this.authService.refreshUserProfile(updatedProfile);
            },
        });
    }

    getMusicPath(musicId: string): string {
        if (musicId === MUSIC_OFF_ID) return '';
        const music = this.musics.find((item) => item.id === musicId);
        return music ? music.path : this.musics[0].path;
    }

    getMusicCover(musicId: string): string {
        const music = this.musics.find((item) => item.id === musicId);
        return music ? music.cover : this.musics[0].cover;
    }

    getOwnedMusics(ownedIds: string[]): Music[] {
        return this.musics.filter((music) => ownedIds.includes(music.id));
    }

    getShopMusics(): Music[] {
        return this.musics.filter((music) => music.id !== MUSIC_DEFAULT_ID);
    }
}
